src/audio/kpb.c

```diff
@@ -0,0 +1,133 @@
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ */
+
+#include <stdint.h>
+#include <uapi/ipc/topology.h>
+#include <sof/ipc.h>
+#include <sof/audio/component.h>
+#include <sof/audio/kpb.h>
+#include <sof/list.h>
+#include <sof/audio/buffer.h>
+
+/**
+ * kpb_new() - create a key phrase buffer component.
+ * @arg1: generic ipc component pointer.
+ *
+ * Return: A pointer to newly created KPB component.
+ */
+static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
+{
+
+	struct comp_dev *dev;
+	struct sof_ipc_comp_kpb *kpb;
+	struct sof_ipc_comp_kpb *ipc_kpb = (struct sof_ipc_comp_kpb *)comp;
+	struct comp_data *cd;
+
+	trace_kpb("kpb_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_kpb->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KPB, ipc_kpb->config);
+		return NULL;
+	}
+
+	/* Validate input parameters */
+	if (ipc_kpb->channels > MAX_SUPPORETED_CHANNELS) {
+		trace_kpb_error("kpb_new() error: "
+		"nr of channels exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->history_depth > MAX_BUFFER_SIZE) {
+		trace_kpb_error("kpb_new() error: "
+		"history depth exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->sampling_freq != MAX_SAMPLNG_FREQUENCY) {
+		trace_kpb_error("kpb_new() error: "
+		"requested sampling frequency not supported");
+	return NULL;
+	}
+
+	if (ipc_kpb->sampling_width != SAMPLING_WIDTH) {
+		trace_kpb_error("kpb_new() error: "
+		"requested sampling width not supported");
+	return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		COMP_SIZE(struct sof_ipc_comp_kpb));
+	if (!dev)
+		return NULL;
+
+	kpb = (struct sof_ipc_comp_kpb *)&dev->comp;
+	memcpy(kpb, ipc_kpb, sizeof(struct sof_ipc_comp_kpb));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	cd->history_depth = ipc_kpb->history_depth;
+	comp_set_drvdata(dev, cd);
```

Comment:
@mrajwa should r_ptr be initialized here?

Reply:
@ranj063, r_ptr of source or internal buffer? The former one is the responsibility of the source itself I think. The later is my job and will be initialized at buffering time.